Feature/typescript

samijaber opened the pull request:
#### Overview

- 🚧 


#### TO-DO
- [x] update README
- [x] complete breaking changes migration guide
- [ ] check what gets exported by the dist & how (including enums such as `Language` and string literal unions for args)
- [x] tests
- [x] look into removing dependency on node built-in `url` module
- [x] document:
  - [x] special response shape for feeds
  - [x] error sources
 - [x] allow empty object for optional params on methods

lukechesser reviewed (commented):
Outside of the methods we don't implement, are there any breaking changes?

samijaber commented on src/methods/photos/index.ts:
```diff
@@ -0,0 +1,56 @@
+    },
+  });
+
+export const track = (photo: { links: { download_location: string } }) => {
```
Yeah, I think it's totally fine. The argument type here is a placeholder, I'm going to change it so it's only expecting a `downloadLocation` string for simplicity's sake. That way we're not expecting a full Photo object which is unreasonable. And then we can always add more args for additional query params as we need.

samijaber replied:
f46144e

samijaber commented:
> Outside of the methods we don't implement, are there any breaking changes?

There's more! It's pretty much all changed. I'll need to write up a new readme of course. Biggest things:
- removed a lot of methods
- renamed all existing methods & changed their arguments
- changed the response type of the Promise for all the methods: 
  - it now automatically calls `response.json()` if the `response` is valid JSON and returns that as its output
  - it does some basic error-handling now.
- added the ability to abort network requests
- the API is an object instead of a class, so creating an instance is no longer done using `new Unsplash()`

A very quick overview of how to consume response from the new API:

```ts
const api = Unsplash({ accessKey: 'abc123' });

api.photos.get({ photoId: 'foo' }).then(result => {
  if (result.type === 'error') {
    console.log('error occurred: ', result.errors[0]);
  } else if (result.type === 'aborted') {
    console.log('fetch request aborted');
  } else {
    console.log('received photo: ', result.response);
  }
});
```

One other thing I am also adding is for feed methods: the lib will automatically handle extracting the total from `x-total` header and return that along with the JSON body (e.g. `result.response.feed` and `result.response.total`) as that's annoying manual work that any Unsplash API consumer that shows a feed will likely need for the sake of pagination.

lukechesser commented:
@samijaber cool, we'll need to create an upgrade guide to go with the new README as well. Something that tells people how to go from the old methods to the new ones, and any 'gotchas'

OliverJAsh commented on src/helpers/request.ts:
```diff
@@ -0,0 +1,89 @@
+        ...generalFetchOptions,
+      })
+        .catch(err => {
+          if (err.name === 'AbortError') {
+            return ABORTED;
+          } else {
+            throw err;
+          }
```
I'm curious why you decided to catch the `AbortError`. It might be more idiomatic/intuitive to consumers if we just mirror the behaviour of `fetch`, which is to throw the error. Of course, if someone chooses to wrap this with an observable, they can just omit the `AbortError` entirely, which is what `Rx.fromFetch` does when it wraps `fetch`: https://github.com/ReactiveX/rxjs/blob/6.5.5/src/internal/observable/dom/fetch.ts#L99

samijaber replied:
That's a good point. This logic is definitely moving away from native spec behaviour. I didn't have `Rx.fromFetch` in mind when I did this, but it seems like I did something similar to them by treating the abort in a special way. My instinct was to avoid requiring consumers to handle the error because frankly, I find it odd that it's an Abort_Error_...there's no error, everything's fine 😅 . But there's not much we can do about that and we should mirror the spec like you said.

Our options:
- do nothing: let `AbortError`s get thrown as they usually would
- catch aborted reqs and return them in a special way like I have done (`result.type === 'aborted'`)
- do not integrate the `AbortController` at all: it would be fairly easy for those who want it to integrate it when they want to.

I'm starting to lean towards the 3rd one to keep this SDK more straightforward and with one less dependency (even though it's fairly small, users would have to add the `abort-controller` polyfill for certain browsers/node etc.). In fact, users who want to have a shared `AbortController` between multiple calls/endpoints will want to be able to override the `signal` config themselves. At that point, it feels best to drop it entirely 🤔 . WDYT?

samijaber replied:
I removed abort controller: https://github.com/unsplash/unsplash-js/pull/143/commits/d26c3db8d540cfbfffcca0965ab005a0ad80cfaa

OliverJAsh replied:
I hadn't actually looked at the way you were using abort controllers/signals. The way I expected this would work is the user would create a controller themselves and pass the signal into the API function call. Is that how it works now?

samijaber replied:
Yeah! That's how it'll work now 👍

samijaber replied:
Yeah, although they can only pass the signal to the overall API object once:

```ts
const controller = new AbortController();
const signal = controller.signal;

const api = createApi({ accessKey: '', signal });
const photosPromise = api.photos.list().catch(err => {
  // handle abort error
});
const searchPromise = api.search.getPhotos({ query: 'dog' }).catch(err => {
  // handle abort error
});

controller.abort(); // will impact all calls
```

Allowing the user to provide a custom `signal` to an individual call/request would entail us updating the signature of every endpoint's `requestParams` to also include `signal`...not sure if that's worth the effort 🤔

OliverJAsh replied:
> Yeah, although they can only pass the signal to the overall API object once:

Ah, that could be quite limiting. Cancellation doesn't really make sense unless you can use it for individual/specific requests as opposed to all requests made by the client.

Maybe that's a good reason to pass the "dependencies" last instead of upfront (going back to this discussion https://github.com/unsplash/unsplash-web/pull/5900#discussion_r522091979).

Alternatively, perhaps we can find a way to merge the "dependencies" (access key and signal) into the request params, so the user only has to call one function:

```ts
const controller = new AbortController();
const signal = controller.signal;

const photosPromise = api.photos.list({ accessKey: '' }).catch(err => {
  // handle abort error
});
const searchPromise = api.search.getPhotos({ query: 'dog', accessKey: '', signal }).catch(err => {
  // handle abort error
});

controller.abort(); // will only abort the search request
```

Without this I don't think it would be possible for someone to wrap API functions in an `Observable` (one which calls `abort` on unsubscribe).

samijaber replied:
You're right!

Having to provide the `accessKey` on every request seems highly unusual to me, so I'd rank this option last in terms of dev ergonomics.

I did some shuffling around so that users can provide an additional argument containing any `fetch` options on a per-call basis!

samijaber commented on src/helpers/request.ts:
```diff
@@ -0,0 +1,96 @@
+  handlers: RequestGenerator<RequestArgs, ResponseType>,
+) => (
+  ...a: Parameters<typeof handlers['handleRequest']>
+) => Promise<ApiResponse<ResponseType>>;
```
🤯

samijaber commented on src/helpers/request.ts:
```diff
@@ -0,0 +1,85 @@
+  accessKey?: string;
+  apiVersion?: string;
+  apiUrl?: string;
+} & OmitStrict<RequestInit, 'method' | 'body'>;
```
Since we want to support both client and server use, there are two options:

- if `apiUrl` is not defined (i.e. we're calling the API directly without a proxy), then `accessKey` _must_ be defined for auth.
- if `apiUrl` is defined, then I _think_ `accessKey` must _not_ be defined since the only reason for defining the URL is to point to a proxy, in which case we want to make sure the user does not provide the `accessKey` by mistake.

We can easily do that by adjusting the types to be a union:

```ts
type InitParams = {
  apiVersion?: string;
} & OmitStrict<RequestInit, 'method' | 'body'> &
  ({ accessKey: string; apiUrl?: never } | { apiUrl: string; accessKey?: never });
```

WDYT? I'm trying to think of a scenario where someone wants to be able to provide both, but I can't think of any logical one. I'm torn between enforcing the API guidelines through the types (`accessKey` must never be exposed) vs allowing for flexibility in some exotic use-case  🤔

OliverJAsh replied:
> I'm trying to think of a scenario where someone wants to be able to provide both, but I can't think of any logical one.

Me neither. I like your idea! Side note: we don't really need the `?: never` properties.

samijaber replied:
Sweet, I'll make the change.

I found that we need the `?: never` for a bunch of reasons:

- without it, TS will still allow the user to provide both, making my change pointless :confused:
```ts
type InitParams = { accessKey: string } | { apiUrl: string };
const foo = (a: InitParams) => a

foo({ apiUrl: '', accessKey: ''}) //no error
```

- without it, setting default properties (which we do for `apiUrl`) and reading the 2 values becomes more laborious, since we can no longer de-structure those two properties from the function argument right away.

lukechesser replied:
Makes sense to me 👍  I'd just make sure we're clear in whatever docs. The whole client vs server thing can be confusing cause most places wouldn't care about the client making calls to the API directly, but we have to due to the access key leaking

OliverJAsh commented on src/methods/photos/index.ts:
```diff
@@ -0,0 +1,77 @@
+};
+
+export const get = {
+  handleRequest: createRequestParams(({ photoId }: PhotoId) => ({
```
Re. https://github.com/unsplash/unsplash-web/pull/5900#discussion_r522105536

Are we able to re-use these `handleRequest` functions in our own client now?

samijaber replied:
Yeah, that's the plan! I wanted to make sure before merging this PR but the refactoring work in our client is a bit too involved due to the new `createRequestParams` signature etc. I decided to merge this and then we'll tackle the changes in the other repo. If anything is missing here we can always make adjustments (no breaking changes since it's all under-the-hood functionality)

OliverJAsh commented on src/helpers/request.ts:
```diff
@@ -0,0 +1,85 @@
+/**
+ * helper used to type-check the arguments, and add default params for all requests
+ */
+export const createRequestParams = <Args extends {}>(fn: (a: Args) => BaseRequestParams) => (
```
`createRequestHandler` might be a better name for this, for consistency with the `HandleRequest` type I suggested in another comment.

OliverJAsh reviewed (commented):
🎉

samijaber commented on src/helpers/query.ts:
```diff
@@ -0,0 +1,15 @@
+  page,
+  perPage,
+  orderBy,
+}: ValidateShape<T, PaginationParams>) => ({
```
The reason I added it here is to avoid a mistake that I only caught by luck. Example:

```ts
export type SearchParams = {
  query: string;
} & PaginationParams;

// no errros
createRequestParams((paginationParams: SearchParams) => ({
  pathname: `${SEARCH_PATH_PREFIX}/collections`,
  query: Query.getFeedParams(paginationParams),
}))

// should have been:
createRequestParams(({ query, ...paginationParams }: SearchParams) => ({
  pathname: `${SEARCH_PATH_PREFIX}/collections`,
  query: { query, ...Query.getFeedParams(paginationParams) },
}))
```

You're right that it should technically be everywhere but that's not plausible. I ended up deciding to keep it where I caught a bug.

OliverJAsh replied:
> I ended up deciding to keep it where I caught a bug.

I feel your pain about TS's lack of excess property checks in some scenarios, but using `ValidateShape` here just feels a bit random to me—other users might make the same mistake when they call other functions, and we have no protection there. For that reason, I'm not sure it's worth keeping this, unless we use it relentlessly.

samijaber replied:
Fair, I'll remove it and swap it out with some tests for the API call arguments

samijaber commented on package.json:
```diff
@@ -37,27 +70,10 @@
-    "url-parse": "1.4.5"
+    "@types/content-type": "^1.1.3",
+    "content-type": "^1.0.4",
+    "url": "^0.11.0",
```
Right! I got mixed up with something else. Thanks for catching this.

From your experience, is it customary to mention our usage of `url` in READMEs, or does webpack's automatic handling of it make it a non-issue for the vast majority of users?

OliverJAsh replied:
> From your experience, is it customary to mention our usage of `url` in READMEs, or does webpack's automatic handling of it make it a non-issue for the vast majority of users?

Hmm, I don't have much experience really, other than we don't mention it in the docs for `url-transformers` but I'm not sure how many people actually use that.

I guess it doesn't hurt to mention it, but we could wait and see.

samijaber commented on src/methods/collections/index.ts:
```diff
@@ -0,0 +1,45 @@
+      query: Query.getFeedParams(paginationParams),
+    }),
+  ),
+  handleResponse: handleFeedResponse<any>(),
```
I think that would be too restrictive for those who migrate to this new version while we're still building out the types. Even if types are technically not breaking changes, TS users whose codebases allowed `any` from our API responses will effectively be unable to use this new version unless they manually write the API response types (which we're doing for them very soon) or cast everything to `any`. I'd rather go with the more permissive approach here. 

As you know from your experience editing express types, the TS world is full of people who haven't fully bought into types 😛 . I'd rather avoid the onslaught of somewhat-misguided GitHub issues about how we broke everything by moving to `unknown` 😂

OliverJAsh replied:
Agreed, good call!

lukechesser commented on src/helpers/feed.ts:
```diff
@@ -0,0 +1,30 @@
+    if (Number.isInteger(total)) {
+      return total;
+    } else {
+      throw new Error('Expected x-total header to be valid integer.');
```
What if the API throws a 503 error? Will this error get returned before the API error? Because it will be confusing to a user if this error gets returned before the 5xx error, as they'll be wrongly debugging the wrong error

samijaber replied:
This function is only reached if `response.ok === true`, i.e. if the status code is in the 2xx range. So it won't override any APi errors.

lukechesser reviewed (approved):
Looks good from the code perspective 👍  The migration guide and stuff will be good to see just to make sure of what we're asking from users upgrading

samijaber commented on README.md:
```diff
@@ -3,175 +3,214 @@
-  timeout: 500 // values set in ms
+To create an instance, simply provide an _Object_ with your `accessKey`.
+
+NOTE: If you're using `unsplash-js` publicly in the browser, you'll need to proxy your requests through your server to sign the requests with the Access Key to abide by the [API Guideline](https://help.unsplash.com/articles/2511245-unsplash-api-guidelines) to keep keys confidential. We provide an `apiUrl` property that lets you do so:
```
Good catch! I checked on Slack and the help article is out of date. Luke will update it with a mention of the Access Key. Thanks. 👍

OliverJAsh replied:
Should we also mention it in the API docs?

samijaber commented on src/helpers/request.ts:
```diff
@@ -0,0 +1,85 @@
+ */
+type InitParams = {
+  accessKey?: string;
+  apiVersion?: string;
```
I asked the API and there's only ever been a v1, and they're not really sure what versioning they'd do if they ever need it. So I'll just keep it as-is for now.

samijaber commented on src/helpers/errors.ts:
```diff
@@ -0,0 +1,24 @@
+import { AnyJson, checkIsString, isDefined } from './typescript';
+
+type ErrorResponse = {
+  errors: string[];
```
> I think this should be non-empty.

Yup! Done. 👍 

> I don't think we ever have more than one error?

I have seen multiple errors in some endpoints (think topic submissions where each submission could error for a different reason). Could be unlikely for the ones currently implemented, but I prefer keeping options open for the future and mirroring as much as possible the way the real API works.

samijaber commented on src/methods/photos/index.ts:
```diff
@@ -0,0 +1,77 @@
+  handleRequest: createRequestParams(({ downloadLocation }: { downloadLocation: string }) => {
+    const { pathname, query } = urlHelpers.parse(downloadLocation, true);
+
+    if (!isDefined(pathname)) {
```
I wouldn't bother too much IMO 🤷

samijaber commented on src/helpers/errors.ts:
```diff
@@ -0,0 +1,24 @@
+
+// https://stackoverflow.com/a/8511332
+const checkHasErrors = (response: AnyJson): response is { errors: AnyJson } =>
+  isDefined(response) && typeof response === 'object' && 'errors' in response;
```
Done in https://github.com/unsplash/unsplash-js/pull/143/commits/c83070e2852f591838d9b7353692a9fe3e76b1ad. LMK what you think!

samijaber commented on src/helpers/feed.ts:
```diff
@@ -0,0 +1,30 @@
+  } else {
+    throw new Error('Expected x-total header to exist.');
+  }
+};
```
https://github.com/unsplash/unsplash-js/pull/143/commits/37b28151c64b83e4def2fc0bbcf6130928bc44d7

samijaber commented on src/helpers/json.ts:
```diff
@@ -0,0 +1,25 @@
+ */
+export const getJsonResponse = (response: Response): Promise<AnyJson> => {
+  if (checkIsJsonResponse(response)) {
+    return response.json();
```
https://github.com/unsplash/unsplash-js/pull/143/commits/37b28151c64b83e4def2fc0bbcf6130928bc44d7

samijaber commented on src/helpers/errors.ts:
```diff
@@ -0,0 +1,24 @@
+  } else {
+    return [
+      'Responded with a status code outside the 2xx range, and the response body is not recognisable.',
+    ];
```
https://github.com/unsplash/unsplash-js/pull/143/commits/2f004d4e117ff28074ac1d3bacaaf42b735e9194

samijaber commented on src/types/request.ts:
```diff
@@ -0,0 +1,15 @@
+  POPULAR = 'popular',
+  OLDEST = 'oldest',
+}
+type Orientation = 'landscape' | 'portrait' | 'squarish';
```
I'm content with unions for now. Can add it to a nice-to-have list for a 2nd pass

samijaber commented on .eslintrc.js:
```diff
@@ -0,0 +1,21 @@
+/** @type {import('eslint').Linter.Config} */
```
I gave it a quick shot but it's not giving me type-checking errors, so I'll drop it for now. Will add it to a nice-to-have list for later though 👍

OliverJAsh reviewed (approved):
Awesome work on this!

samijaber commented on tests/index.test.ts:
```diff
@@ -44,24 +45,52 @@ const paramsTests: Record<Section, { [index: string]: CompleteRequestParams[] }>
     list: [
       photos.list.handleRequest({ orderBy: OrderBy.LATEST, page: 4, perPage: 10 }),
       photos.list.handleRequest({ orderBy: OrderBy.LATEST }),
+      // TO-DO: figure out type arthimetic to allow for no object to be provided
+      photos.list.handleRequest({}),
```
Yeah! I went down a TS rabbit-hole and couldn't get it quite to work. I'm kinda pissed haha, might spend a bit more time trying to get it working even if it means duplicating the method factory.

samijaber replied:
I worked around this to still provide users the handiness of not providing an empty object: https://github.com/unsplash/unsplash-js/pull/143/commits/81705fb0438879a17eae91f89713d351d034af5c

samijaber commented on package.json:
```diff
@@ -2,22 +2,59 @@
+    "size": "size-limit",
+    "analyze": "size-limit --why"
+  },
+  "resolutions": {
```
Yeah, I'm with you. It's fallen a bit behind from TS releases. WIll keep an eye out and likely eject the next time we need a new version of something and `tsdx` is too far behind.

samijaber commented on src/helpers/request.ts:
```diff
@@ -43,7 +43,7 @@ export const createRequestHandler = <Args>(
   return {
     ...baseReqParams,
     ...additionalFetchOptions,
-    queryToUse,
```
Good point! WIll apply your changes, thanks 👍

samijaber commented on src/helpers/request.ts:
```diff
@@ -0,0 +1,100 @@
+
+type RequestGenerator<Args, ResponseType> = {
+  handleRequest: HandleRequest<Args>;
+  handleResponse: HandleResponse<ResponseType>;
```
By getting rid of user authentication from the lib, it should be impossible to get a 201s (user auth is needed for all endpoints that create entities). All the current endpoints on here should only be `GET`s that return non-empty JSONs with 200s.

If we ever bring back user auth, then we'll def need to better handle responses like you said.

samijaber commented on CHANGELOG.md:
```diff
@@ -1,10 +1,86 @@
 # Changelog
 
+## 7.0.0
```
@lukechesser I added a changelog for v7 and factored in all your feedback. Since the breaking changes are fairly substantial, I often pointed to the updated README.

Let me know if you see anything in particular that's missing.

lukechesser replied:
Perfect — this looks great to me 👍  Straightforward

lukechesser commented on src/index.ts:
```diff
@@ -0,0 +1,39 @@
+  photoId,
+}: {
+  photoId: string;
+}) => fetch(`views.unsplash.com/v?photo_id=${photoId}&app_id=${appId}`);
```
nice! Thanks for adding this. That's fine that it appears in the types.

We support a few more options for it — it would be convenient if we could support it without a ton of work. I'll send you the private docs on Slack

OliverJAsh commented on src/helpers/url.ts:
```diff
@@ -0,0 +1,48 @@
+/**
+ * We currently avoid relying on a large library for URL parsing/building to keep bundle size small,
+ * and avoid env-specific dependencies such as node's `url` or `URL` (which is not available in IE11).
+ *
+ * The helpers here do the bare minimum needed to handle use-cases in this library, and might not
+ * handle more complex scenarios with edge-cases.
```
I'm not sure I understand how this is better (using a third party lib instead of a Node built-in modules such as `querystring` or `url` which are pretty much global because of shimming). Surely the consumer's bundle will be bigger if they have to bundle both `url`/`querystring` (I suspect most people are already bundling these as they're used almost universally) _and_ this module's dependency on a third part lib `querystringify`. 🤔

If there's a reason not to use `url` specifically, we could at least use the built-in `querystring` module instead of `querystringify`, so there's less chance users will end up bundling multiple dependencies which do the same thing.

OliverJAsh replied:
Alternatively we could just use `URL`/`URLSearchParams` and ask users to polyfill if they need IE11 support.

samijaber commented on src/helpers/url.ts:
```diff
@@ -0,0 +1,48 @@
+ *
+ * The helpers here do the bare minimum needed to handle use-cases in this library, and might not
+ * handle more complex scenarios with edge-cases.
+ */
```
@OliverJAsh Bear with me 😂 

The reason I went with this route is that I wasn't comfortable shipping a JS SDK that depends on a NodeJS built-in. I didn't want to ship an SDK with a caveat that says "if you're using webpack, you're _most_ likely fine but if you use anything else you'll need a polyfill to use this in the client". I want it to be easily used by everyone, and having to worry about node-js built-ins polyfills could be very daunting for a beginner. 

From there, my options were:
- find an env-agnostic module to do it all
- use `URL` which is supported in both [browsers](https://developer.mozilla.org/en-US/docs/Web/API/URL) and [node](https://nodejs.org/api/url.html)
- write it myself

1- From my searches, I couldn't find an env-agnostic module that was small enough in size.
2- URL seemed promising but it has no IE11 support (not sure how big of an issue that is, but it would be nice to support as much as possible?), but also it meant writing code to identify being in Node VS browser, which meant also choosing whether to handle web workers or not.

I ended up choosing to delegate query logic to a very small lib `querystringify` which handles our needs. It's not comprehensive but is much smaller than alternatives. I then wrote functions to replace `url` and `url-transformers` to cover our use-cases.

I think this approach is safe and a good trade-off. I'm curious what you think though and if you have strong opinions about doing it differently. I'm also open to attempting using `URL`, but once I got it done this way and added tests I figured it's in a good place. WDYT?

samijaber replied:
replying to your other comment here:

I guess I was worried about how safe it is to rely on nodejs built-ins shimming. 

I see your point about us increasing the overall size of users' bundles by recreating code they're most likely already shimming due to all of the other packages they have in their code! Maybe I don't need to worry about any of this and we can just ship it, wait and see if anybody complains!

OliverJAsh replied:
> URL seemed promising

> it meant writing code to identify being in Node VS browser, which meant also choosing whether to handle web workers or not.

Interesting, how come it meant you had to do this? I thought we would just use the global classes in both environments. We wouldn't need any conditional logic, or would we?

I think that would be my favourite approach. Use the modern APIs which are available globally and ask users to polyfill if they need support in old browsers.

samijaber replied:
> We wouldn't need any conditional logic, or would we?

I initially thought we needed to access it on node using `require('url').URL`. Didn't realize it was also available on the global scope! 🤦 

I've reverted my latest changes for now. Will give `url` vs `URL` some more thought

samijaber replied:
Went with `URL`!

OliverJAsh commented on src/helpers/url.ts:
```diff
@@ -0,0 +1,35 @@
+  query: Query;
+};
+
+const addQueryToUrl = (query: Query) => (url: URL) => {
+  Object.keys(query).forEach(queryKey =>
+    url.searchParams.set(queryKey, query[queryKey].toString()),
+  );
+
+  return url;
+};
+
+const createUrl = (pathname: string) => (apiUrlBase: string) => new URL(pathname, apiUrlBase);
+
+export const buildUrl = ({ pathname, query }: BuildUrlParams) =>
+  flow(createUrl(pathname), addQueryToUrl(query), url => url.toString());
```
We don't technically need to return anything in `addQueryToUrl`, since this is just mutating the `URL` that was passed in. It doesn't hurt to return it anyway, but I think if we're using a mutable API, we should make that explicit as possible. `flow` is best suited to immutable functions, so I'm not sure we should use it here.

```suggestion
const addQueryToUrl = (query: Query) => (url: URL) => {
  Object.keys(query).forEach(queryKey =>
    url.searchParams.set(queryKey, query[queryKey].toString()),
  );
};

export const buildUrl = ({ pathname, query }: BuildUrlParams) => (apiBaseUrl: string) => {
  const url = new URL(pathname, apiBaseUrl);
  addQueryToUrl(query)(url);
  return url.toString();
};
```

OliverJAsh commented on src/helpers/url.ts:
```diff
@@ -0,0 +1,35 @@
+import { flow } from './fp';
+
+type Query = {
+  [index: string]: string | number | boolean;
+};
```
Could we replace all instances of this with `URLSearchParams`?

samijaber replied:
Not quite, because URLSearchParams is not a plain object but a class instance. So we'd have to upgrade our `query` arguments created in our methods to actually be `new URLSearchParams(query)`. It's a bit too involved of a change so I'd rather leave that out for later.

OliverJAsh replied:
Yeah, I was suggesting using `URLSearchParams` everywhere, rather than try to maintain two representations of the same thing.

samijaber replied:
I see. I've added it to https://github.com/unsplash/unsplash-js/issues/145 to tackle post-adding the response types.

OliverJAsh commented on src/methods/collections/index.ts:
```diff
@@ -0,0 +1,52 @@
+      ...paginationParams
+    }: CollectionId & PaginationParams & OrientationParam) => ({
+      pathname: `${COLLECTIONS_PATH_PREFIX}/${collectionId}/photos`,
+      query: compactDefined({ ...Query.getFeedParams(paginationParams), orientation }),
```
Do we need to use `compactDefined` here? I thought you decided to do that inside `createRequestHandler`?

Also applies to other places in this PR.

samijaber replied:
Oh sorry. I ended up tightening the `Query` type to not allow any non-values after all, which required me to add `compactDefined` wherever it was needed in methods. We no longer need to compact in `createRequestHandler`, so I removed it from there just now.
